### ISSUE TITLE
auro-button: fluid button displays incorrectly in Safari when built in a Svelte app

### ISSUE THREAD

**Issue by CarsonTolleshaug** [AlaskaAirlines/auro-button#119]
## Describe the bug
Button appears to display as text when rendered in Safari as part of a Svelte app and given the "fluid" attribute. 

## To Reproduce
1. Include `<auro-button fluid type="submit" >Button</auro-button>`  on page in a svelte 3.38.2 app.
2. Launch page in Safari browser.

## Expected behavior
Button should display with blue background, rounded edges, and should fill the width of the parent container.

## Screenshots
In Chrome:
![button-chrome](https://user-images.githubusercontent.com/11671889/119414242-0e916480-bca4-11eb-9fd4-766c6bd638f3.png)

In Safari:
![button-safari](https://user-images.githubusercontent.com/11671889/119414266-181acc80-bca4-11eb-9e61-1095420062bd.png)

## Desktop (please complete the following information):
 - OS: OSX
 - Browser: Safari
 - Version: 6.2.0

## Additional context
I don't have a mac so I have no way of testing myself which parts of this may be causing the problem. I don't know for sure if the "fluid" part has any effect on the bug.
Issue originally reported by Emiliia Kostiukevych, please include her in any correspondence.

Button displays correctly in other browsers.

**Comment by geoffrich** [comment]
This isn't a Svelte issue, but a legacy browser issue. Custom Elements are only supported by [Safari 10](https://caniuse.com/custom-elementsv1) and later. I don't think we currently support Safari 6, since it's quite old at this point. If you do need to support Safari 6, you'll need to [polyfill](https://github.com/AlaskaAirlines/auro/blob/master/src/POLYFILL.md) and transpile the ES6 code using Babel.

@blackfalcon can likely shed more light on Auro's current browser support strategy.

**Comment by geoffrich** [comment]
Though now I'm confused why the checkbox and input fields are displaying correctly... are you able to provide a demo page that repros the issue?

I created a [Svelte REPL](https://svelte.dev/repl/26c38802d3764726ae9cb864beaa1f0b?version=3.38.2) that I tested on the latest iOS safari and all works as expected.

**Comment by blackfalcon** [comment]
@CarsonTolleshaug what do you see with https://auro.alaskaair.com/components/auro/button

Also, are you sure you are on Safari 6? None of this should work? Are you using the CDN versions? 

I agree with @geoffrich we are not supporting anything more than a [current release](https://auro.alaskaair.com/support/browsersSupport) really.

**Comment by blackfalcon** [comment]
oops

**Comment by blackfalcon** [comment]
> I created a Svelte REPL that I tested on the latest iOS safari and all works as expected.

![](https://media.giphy.com/media/4SQMqhWzUA0Fi/giphy.gif)

**Comment by blackfalcon** [comment]
@CarsonTolleshaug is this still an issue for you?

**Comment by blackfalcon** [comment]
@CarsonTolleshaug is this still an issue? We are unable to repro the issue. Unless there is a repro we can see, this issue will be closed.

**Comment by diana-wang** [comment]
We're still seeing this issue, though it is definitely a browser-based issue, not something specific to Svelte. Here is an example in Flight Search's Svelte microsite: www.alaskaair.com/search/sea/avp/26sep. The "Add to cart" text should be a button

![image](https://user-images.githubusercontent.com/11242098/122988432-99728700-d356-11eb-8202-015e1dd224f0.png)

And here's a screenshot from our React seat map application, which is using Component Hanger to serve auro files.

![image](https://user-images.githubusercontent.com/11242098/122987833-e99d1980-d355-11eb-9364-9bedf5f5ffd3.png)

Component hanger is currently using 6.2.0, so I tried updating Component Hanger to use the auro-button 6.3.3, but that broke the headless chrome tests, which I'm attempting to fix now ([pipeline](https://dev.azure.com/itsals/E_Sell_PathtoPurchase/_build/results?buildId=1745886&view=results)). Running locally, this is what I see:

![image](https://user-images.githubusercontent.com/11242098/122988117-3aad0d80-d356-11eb-8d12-ce29f3a4acb6.png)

Any help on addressing this would be great! Please let me know if I should create a separate issue for this.

**Comment by geoffrich** [comment]
@diana-wang I can't repro the button issues you describe above in Chrome or Firefox (see screenshots below). What browsers are you using?

![image](https://user-images.githubusercontent.com/4992896/123001377-81eeca80-d365-11eb-8614-d34a41bf5857.png)
![image](https://user-images.githubusercontent.com/4992896/123001426-929f4080-d365-11eb-93c9-8072ee69a4ea.png)

**Comment by diana-wang** [comment]
@geoffrich Safari is the only one that's broken on desktop. Chrome and Firefox are both fine.

Though I think your PR #128 might fix this issue... I updated the lit-element and lit-html version like you suggested in #127 and the buttons are working now, at least in the seat map. 

![image](https://user-images.githubusercontent.com/11242098/123014143-6b06a300-d37a-11eb-932f-6a8c384a8d03.png)

**Comment by liambrozik** [comment]
I am having a very similar issue with auro-button version 6.3.4 in a React app not properly rendering in Safari version 14.1.1 on desktop and also Safari mobile (iOS 14.4.2).

You can see it in this screenshot (where there are a couple other auro components rendering properly):
<img width="1317" alt="Screen Shot 2021-06-24 at 10 51 38 AM" src="https://user-images.githubusercontent.com/17190016/123311001-84733080-d4db-11eb-8238-fbd1068ad86e.png">

**Comment by CarsonTolleshaug** [comment]
@blackfalcon Sorry for not responding. I don't have any way of testing this, as I don't have any Apple products (Safari can't be run on non-apple products). I made this post on behalf of Emiliia Kostiukevych. She is our Designer and unfortunately doesn't have a GitHub account. I can ask her to test, but honestly it would probably shorten your feedback loop if you talk to her directly.

Or it also seems like some others have joined the chat that may be better test subjects anyway.

Sorry for not being very helpful. I wish I could test this myself :(

**Comment by geoffrich** [comment]
I just paired with @liambrozik and we've figured out what's going on in his app (trips). I think the cause is similar in the apps above. 

The root cause of the issue is outlined in #112. This was introduced in 6.2.0 and fixed in 6.3.1. I was able to repro the issue with those versions in AuroReactDemo, and the issue is not present in the latest (6.3.4). Installing the latest version of the button _should_ fix the issue.

However, there is a custom element quirk at play here. Custom elements are different than React or other UI components. There can only be one definition of a custom element on a page at a given time. So, once `customElements.define('auro-button, AuroButton')` is called, it can't be re-defined -- the first one to define the custom element wins out.

This becomes an issue when multiple scripts are on the page each trying to load their own version of the Auro components. Trips was loading a component from the P2P CDN (component hangar), which imports auro-button v6.2. Because the P2P script was loaded first, auro-button v6.2 was the version used despite the bundled app importing auro-button v.6.3.4.

We were able to work around it in trips by adding the "defer" attribute to the P2P scripts, so that they are loaded in the document last. This means the P2P components will be using the latest version of auro-button. In this case it didn't matter, but it could in the future if a page is written to use v1 of a component and the imported P2P component uses v2 (or vice versa);

tl;dr
1. If you are not loading components from the P2P CDN, update your local auro-button to v6.3.4
2. If you are loading components from the P2P CDN, wait for the P2P CDN to use the latest version OR update on your end and load their script tags last.

@diana-wang @CarsonTolleshaug does this sound like it will fix the issue in your apps?

@blackfalcon I think we need to figure out a strategy for allowing multiple versions of auro components to exist on a page. Not sure what that looks like.

**Comment by blackfalcon** [comment]
@geoffrich yup... I've been waiting for this issue to show up. Super difficult to defensively code for, but now that we are here it's time to talk about this. 

This is something where we need to reach out to the greater WC community.

**Comment by blackfalcon** [comment]
What's interesting here is why it breaks in Safari, but not Chrome? Guess? Webkit is behind the 8-ball when it comes to CE support. I feel like Chrome is applying some magic here. 

What can we do to create a demo that illustrates the issue so that we can inquire into the community?

**Comment by geoffrich** [comment]
@blackfalcon I think you might be conflating two separate issues here.

1. The buttons broke in Safari because of #112, because we were applying an attribute illegally in the constructor with the focus visible polyfill. At the time we only thought this broke tests, but it seems like it broke Safari as well. It wasn't an issue in Chrome because we only applied the attribute if the browser did not support focus visible, and Chrome supports focus visible. Chrome would have thrown the same exception if we didn't have the focus visible condition, since adding an attribute in the constructor is a violation of the custom element spec. We fixed this issue in 6.3.1 so everything should be good. However...
2. The above issue is exacerbated by the fact that you can only load one custom element definition for a given HTML tag. Because the P2P CDN is loading the broken auro-button, buttons on the page will break in Safari even if the individual app is loading the latest version of auro button. Whoever loads the definition first wins. This is something that [Scoped Custom Element Registries](https://open-wc.org/docs/development/scoped-elements/#motivation) may help with, but we'd have to polyfill that. It's not implemented in any browser yet.

Assuming I'm right about the root cause, the temporary solution is updating the button version on the P2P CDN and/or loading the P2P scripts last so that the app's version of the button wins out. I'll go ahead and create a separate issue on the Generator to discuss strategies for conflicting custom element definitions.

**Comment by blackfalcon** [comment]
The scope of this issue has taken a lot of twists and turns. I am electing to close this issue in favor of https://github.com/AlaskaAirlines/WC-Generator/issues/159 as the core issue is not with auro-button, but this can happen with any CE in the library.
